[MM-13955] Add telemetry for Android

saturninoabril opened the pull request:
#### Summary
This PR adds telemetry for Android.  Telemetry will be configured on Android beta build only [[1](https://github.com/mattermost/mattermost-mobile/pull/2714)] and will send data whenever the server it's connected to has [diagnostics enabled](https://docs.mattermost.com/administration/config-settings.html#enable-diagnostics-and-error-reporting).

Initial metrics to capture are the following:
  - `start:overall`
  - `start:process_packages`
  - `start:content_appeared`
  - `start:select_server_screen`
  - `start:channel_screen`
  - `team:switch`
  - `channel:loading`
  - `channel:switch_loaded`
  - `channel:switch_initial`
  - `channel:close_drawer`
  - `channel:open_drawer`
  - `posts:loading`
  - `posts:list_update`
(Details will be documented per [MM-15171](https://mattermost.atlassian.net/browse/MM-15171).)

The device information to be collected are the following:
  - `apiLevel: DeviceInfo.getAPILevel()`
  - `buildNumber: DeviceInfo.getBuildNumber()`
  - `bundleId: DeviceInfo.getBundleId()`
  - `brand: DeviceInfo.getBrand()`
  - `country: DeviceInfo.getDeviceCountry()`
  - `deviceId: DeviceInfo.getDeviceId()`
  - `deviceLocale: DeviceInfo.getDeviceLocale()`
  - `deviceType: DeviceInfo.getDeviceType()`
  - `deviceUniqueId: DeviceInfo.getUniqueID()`
  - `height: Dimensions`
  - `isEmulator: DeviceInfo.isEmulator()`
  - `isTablet: DeviceInfo.isTablet()`
  - `manufacturer: DeviceInfo.getManufacturer()`
  - `maxMemory: DeviceInfo.getMaxMemory()`
  - `model: DeviceInfo.getModel()`
  - `serverVersion`
  - `systemName: DeviceInfo.getSystemName()`
  - `systemVersion: DeviceInfo.getSystemVersion()`
  - `timezone: DeviceInfo.getTimezone()`
  - `version: DeviceInfo.getVersion()`
  - `width: Dimensions`
(See [react-native-device-info docs](https://github.com/mattermost/react-native-device-info#react-native-device-info) for details on each info.)

@esethna I believe there's no identifiable info included on the list above. But I'd like to request for your review if it conforms with our [Telemetry docs](https://docs.mattermost.com/administration/telemetry.html).

#### Ticket Link
Jira ticket: [MM-13955](https://mattermost.atlassian.net/browse/MM-13955)

#### Device Information
This PR was tested on: [Android and also on iOS simulator (just to check if no side-effect)]

saturninoabril commented on app/components/post_body/post_body.js:
```diff
@@ -104,6 +118,10 @@ export default class PostBody extends PureComponent {
         }
+
+        if (this.props.isLastPost) {
+            this.logTelemetry();
```
Though last post doesn't mean last to render, this makes a good initial indicator on when to end capturing of metrics like overall load time.

saturninoabril commented on app/components/sidebars/main/index.js:
```diff
@@ -15,11 +16,19 @@ import {isLandscape, isTablet, getDimensions} from 'app/selectors/device';
 
     return {
         ...getDimensions(state),
+        loadedChannelIds: getChannelIdsInPostsInChannel(state),
```
I added this props to easily identify whether a channel has been previously loaded or not, knowing that previously loaded channel will render faster than the one which will still going to request from the server.  Just a good metrics to compare.

enahum commented on app/components/post_list/post_list_base.js:
```diff
@@ -72,6 +80,10 @@ export default class PostListBase extends PureComponent {
             this.props.actions.setDeepLinkURL('');
         }
+
+        this.lastPostIndex = this.getLastPostIndex(this.props.postIds);
```
Why getting this value on willMount and didMount?

saturninoabril replied:
It's not necessary.  That was part of my initial code . Thanks for catching.

enahum commented on app/components/sidebars/main/index.js:
```diff
@@ -15,11 +16,19 @@ import {isLandscape, isTablet, getDimensions} from 'app/selectors/device';
 
 import MainSidebar from './main_sidebar.js';
 
+export const getChannelIdsInPostsInChannel = createIdsSelector(
```
Perhaps a better name `getChannelIdsWithPosts`?

saturninoabril commented on android/app/src/main/java/com/mattermost/rnbeta/MainApplication.java:
```diff
@@ -86,10 +101,16 @@ public boolean isDebug() {
             new RNCWebViewPackage(),
-            new RNGestureHandlerPackage()
+            new RNGestureHandlerPackage(),
+            new StartTimePackage(this)
```
Done, thanks

saturninoabril commented on android/app/src/main/java/com/mattermost/rnbeta/MainApplication.java:
```diff
@@ -111,6 +132,8 @@ public void onCreate() {
 
     SoLoader.init(this, /* native exopackage */ false);
+
+    addReactMarkerListener();
```
Done, thanks

saturninoabril commented:
Thanks @enahum for the reviews.
- done including telemetry js to bundle
- only load/listen to native code if telemetry is enabled (not sure if commenting/uncommenting code is acceptable but it works)
- updated fastlane to run telemetry config for Android build only (which is initially implemented)
- addressed other comments

Please take a look again.

enahum commented on app/telemetry/telemetry.android.js:
```diff
@@ -0,0 +1,166 @@
+
+    canSendTelemetry() {
+        const {config} = store.getState().entities.general;
+        return Boolean(!__DEV__ && config.EnableDiagnostics === 'true' && LocalConfig.TelemetryEnabled);
```
Do we want to include that the native module is available as well?

saturninoabril replied:
For me, it's no need. Since we're updating telemetry config via build, I think check to `LocalConfig.TelemetryEnabled` is sufficient enough.

enahum commented on fastlane/Fastfile:
```diff
@@ -383,6 +377,34 @@ platform :android do
 
+  desc 'Update config and module/listener for Android telemetry'
+  lane :configure_telemetry_android do
+    if ENV['BETA_BUILD'] == 'true' && ENV['TELEMETRY_ENABLED'] == 'true' && ENV['TELEMETRY_URL'] != '' && ENV['TELEMETRY_API_KEY'] != ''
```
By removing the check on `BETA_BUILD` we can only include telemetry if the others are set and that could help if we deploy this later on to production without any code changes.

saturninoabril replied:
I agree, now updated. Thanks

hmhealey commented on android/app/src/main/java/com/mattermost/rnbeta/MainApplication.java:
```diff
@@ -45,24 +45,40 @@
   public Boolean sharedExtensionIsOpened = false;
   public Boolean replyFromPushNotification = false;
 
+  public static long APP_START_TIME;
```
Do these need to be static? I assume we'll only have one instance of MainApplication at a time, but it seems like they could just be fields on the object.

saturninoabril replied:
Updated as field.  I did those as static since it's used by another package (StartTime) which now I removed 'cause it's actually not needed.

hmhealey commented on android/app/src/main/java/com/mattermost/rnbeta/MainApplication.java:
```diff
@@ -88,6 +104,16 @@ public boolean isDebug() {
     );
+
+    // Uncomment to load module for build that has telemetry enabled
+    // packages.add(new StartTimePackage(this));
```
Are these lines going to be uncommented as part of our build process?

saturninoabril replied:
Yes, although this particular line is now removed.
It's still the way it is with `// addReactMarkerListener();`

hmhealey commented on app/components/post_body/post_body.js:
```diff
@@ -104,6 +118,10 @@ export default class PostBody extends PureComponent {
         }
+
+        if (this.props.isLastPost) {
+            this.logTelemetry();
```
I don't like doing this in the Post component since those can show up outside of the main view. It definitely doesn't belong in `measurePost` since that should just be for actually measuring the post.

Perhaps we could find some way to do this in the Channel screen's `componentDidUpdate` or `componentDidMount'`? That's around where I'd expect to find it.

saturninoabril replied:
That (channel screen) was the initial location however it's giving shorter time period which is not realistic.  Parent component may already be rendered but the grandchildren are not yet.
In this place is where I'm getting a closer time period to end rendering.  The idea is to place this on the leaf component as much as possible where it can measure activities like image loading, etc.

hmhealey replied:
That makes sense. I'm worried that it'll make it hard to isolate this logic, but it's probably fine.

hmhealey commented on app/components/post_list/post_list_base.js:
```diff
@@ -72,6 +79,10 @@ export default class PostListBase extends PureComponent {
             this.props.actions.setDeepLinkURL('');
         }
+
+        this.lastPostIndex = this.getLastPostIndex(this.props.postIds);
```
Since this is being used in render, it should be in `this.state`. If you moved it to `getDerivedStateFromProps`, it would work about the same as this.

hmhealey commented on app/components/sidebars/main/main_sidebar.js:
```diff
@@ -35,6 +37,7 @@ export default class ChannelSidebar extends Component {
             setChannelLoading: PropTypes.func.isRequired,
         }).isRequired,
         blurPostTextBox: PropTypes.func.isRequired,
+        channelIdsWithPosts: PropTypes.array,
```
This is going to cause the sidebar to re-render a lot when it doesn't need to. Maybe we can make a view action to do this without the extra prop? If you make something like this, you can pass it into the MainSidebar's mapDispatchToProps, it won't re-render every time we load more posts.
```javascript
export function logChannelSwtich(channelId, currentChannelId) {
    return (dispatch, getState) => {
        if (channelId === currentChannelId) {
            return;
        }

        const metrics = [];
        if (getState().entities.channels.postsInChannel[channelId]) {
            metrics.push('channel:switch_loaded');
        } else {
            metrics.push('channel:switch_initial');
        }

        telemtry.reset();
        telemetry.start(metrics);
    };
}
````

saturninoabril replied:
Got it, now updated using this.

saturninoabril commented:
@hmhealey @enahum PR updated, please take a look again, thanks!

enahum commented:
Will do as soon as I fix master

enahum commented on app/components/post_list/post_list.android.js:
```diff
@@ -20,20 +21,28 @@ export default class PostList extends PostListBase {
-            dataSource: new DataSource(props.postIds, this.keyExtractor),
+            dataSource: new DataSource(props.postIds, keyExtractor),
+            lastPostIndex: getLastPostIndex(this.props.postIds),
+            postIds: this.props.postIds,
```
@saturninoabril in this PR #2761 we are removing this file, how do you want to handle it?

saturninoabril replied:
PR rebase and updated.  What I did was instead of deriving a state, I just pass `lastPostIndex` as props into it.  
To apply it for Android only, I made it like,
```javascript
<PostList
    lastPostIndex={Platform.OS === 'android' ? getLastPostIndex(visiblePostIds) : -1}
    ...
/>
```

enahum commented on app/components/post_list/post_list_utils.js:
```diff
@@ -0,0 +1,22 @@
+
+import {isDateLine, isStartOfNewMessages} from 'mattermost-redux/utils/post_list';
+
+export function getLastPostIndex(postIds) {
```
I think we should move this functions to `mattermost-redux/utils/post_list`

saturninoabril replied:
Done.

enahum commented on app/components/post_list/post_list_utils.js:
```diff
@@ -0,0 +1,22 @@
+    return index;
+}
+
+export function keyExtractor(item) {
```
After #2761 this function won't have any sense to have it here

saturninoabril commented:
@enahum PR updated. Please take a look again.  Thanks

enahum commented:
@saturninoabril you still need to merge the redux pr and add the reference to the commit

saturninoabril commented:
@enahum That's right.  Will just wait for last review of it.

enahum commented:
@saturninoabril redux PR merged, feel free to update the ref and merge this one

saturninoabril commented:
PR updated with corresponding commit hash of mattermost-redux.

enahum commented on app/components/post_list/post_list.js:
```diff
@@ -48,6 +49,7 @@ export default class PostList extends PureComponent {
         highlightPostId: PropTypes.string,
         initialIndex: PropTypes.number,
         isSearchResult: PropTypes.bool,
+        lastPostIndex: PropTypes.number.isRequired,
```
@saturninoabril we forgot about the Thread view and permalink, so those screens are showing a warning where `lastPostIndex` is **undefined**, can we make sure we fix that?

saturninoabril replied:
PR submitted - https://github.com/mattermost/mattermost-mobile/pull/2780
